Implement FAISS Indexing and Dimensionality Reduction

adaorachi opened the pull request:
### Description:

This pull request introduces functionalities for creating a FAISS index, populating it with datasets embedding, and implementing UMAP dimensionality reduction techniques to visualize the vectors. The generated data from the dimensionality reduction process is saved in a file-based database.

### Features Implemented

- Created a module for initializing a FAISS index.
- Implemented necessary functions to set up the index with desired parameters.
- Included functions to generate vectors from datasets with specified dimensions and characteristics using imagebind.
- Implemented either UMAP algorithm for dimensionality reduction of the dataset embedding.
- Included functionality to save the transformed data from UMAP into a file-based database.

Kindly review the changes as I look forward to your feedback.

adaorachi committed:
implement load_dataset func

adaorachi committed:
generate embeddings with imagebind

adaorachi committed:
reduce dataset dims using umap

adaorachi committed:
create_faiss_index

adaorachi committed:
cleanup main.py

adaorachi committed:
fix lint errors

adaorachi committed:
fix black lint errors

Mortimerp9 commented on backend/multimodalexplorer/functions/create_faiss_index.py:
```diff
@@ -0,0 +1,142 @@
+# Determine the device (CPU or GPU) for torch operations
+DEVICE = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
+
+# Instantiate imagebind model
+model = imagebind_model.imagebind_huge(pretrained=True)
+model.eval()
+model.to(DEVICE)
```
this will load the model on import, which is quite expensive and might not be what you want to do, I suggest having this in a api call (that checks if the model was already initialized before loading it again). Then you load this when needed in the lifestyle events of fastapi, see example: https://fastapi.tiangolo.com/advanced/events/#lifespan

Mortimerp9 replied:
actually, I guess that you don't need this loaded in the server, just when preprocessing the dataset. But still, it's easier to have this in a function and you call it in your processing script.

Mortimerp9 commented on backend/multimodalexplorer/functions/create_faiss_index.py:
```diff
@@ -0,0 +1,142 @@
+    faiss.normalize_L2(data)
+
+    # Training index
+    index.train(data)
```
you should only train on a subset of the data, try getting a random subset ` numpy.random.choice` with ` replacement=False` should do the trick.

Mortimerp9 commented on backend/multimodalexplorer/functions/helpers.py:
```diff
@@ -0,0 +1,22 @@
+def save_data_as_file(data, file_name, folder="../db"):
+    folder_path = os.path.join(os.path.dirname(__file__), folder)
+    os.makedirs(folder_path, exist_ok=True)
+    csv_path = os.path.join(folder_path, f"{file_name}.tsv")
+
+    df = pd.DataFrame(data)
+
+    df.to_csv(csv_path, sep="\t", index=False)
```
not totally sure why you go back to pandas to save to file. The numpy embeddings should saved with numpy.save/load. For umap, you get a bunch of 2 dimentional points, you could also just dump this with numpy. I see that you try to keep more columns with dataset names etc. In this case saving to CSV might be useful.
For faiss index, you should not save it as a dataframe, faiss has a save API: ` faiss.write_index(returned_index, index_output_file)` and ` faiss.read_index(filename)`.

Mortimerp9 replied:
ok, I looked at the umap API a bit more in details, `fit_transform` returns the whole array of points. If you want to save this as csv in a useful way, you will have to unflatten the points probably.

Mortimerp9 commented on backend/multimodalexplorer/main.py:
```diff
@@ -29,4 +30,12 @@ async def root():
     return {"message": "Welcome to MultiModalExplorer."}
 
 
+def main():
```
https://github.com/gwenzek/func_argparse is a good lib to parse cli arguments in python so you can pass some parameters here.

Mortimerp9 replied:
I don't think that preprocessing should live in the server main. I suggest that you have:
1. a preprocessing script that takes a (set of) dataset as input and creates file in a pre-determined folder structure
2. a server script that can load the preprocessed data. This could be static at the beginning, where it the server startup script takes as a parameter the name of the folder where to find the preprocessed file for a (set of) datasets we want to look at.

Mortimerp9 commented on backend/multimodalexplorer/functions/create_faiss_index.py:
```diff
@@ -0,0 +1,142 @@
+        {"data": data, "media_type": data_type, "embedding": embedding}
+        for data_type, dataset in all_datasets.items()
+        for data, embedding in zip(dataset, embeddings_umap)
+    ]
```
I don't think that this will scale to a lot of data as you are trying to fit a lot of datasets and embeddings in memory.

Mortimerp9 commented on backend/multimodalexplorer/functions/create_faiss_index.py:
```diff
@@ -0,0 +1,142 @@
+    text_dataset = load_datasets()
+
+    # Dictionary containing all datasets
+    all_datasets = {
+        "text": text_dataset,
+    }
```
this doesn't look like a very useful data structure, what if we have more than one text dataset?

If you want to play around with image data, check one of the datasets in https://huggingface.co/datasets?task_categories=task_categories:image-classification&license=license:mit&sort=trending

Mortimerp9 commented on backend/multimodalexplorer/functions/create_faiss_index.py:
```diff
@@ -0,0 +1,142 @@
+    trained_dataset = loaded_dataset["train"]
+
+    # Extract the sentences from each entry in the training dataset
+    dataset = [entry["set"][0] for entry in trained_dataset]
```
each dataset has a slightly different way of storing data sadly, so maybe make this specific to the dataset name.

Mortimerp9 commented on backend/multimodalexplorer/functions/create_faiss_index.py:
```diff
@@ -0,0 +1,142 @@
+
+    # Compute embeddings for each modality
+    with torch.no_grad():
+        embeddings = model(dataset_input)
```
imagebind is good to embed images, sounds and text together, but SONAR (the example in colab) is good for embedding speech and text together. I would love it if we can have a preprocess that can generate the correct data for the UI for both of these embedding models. Once you have the embeddings, the umap and faiss code are exactly the same (except for embedding dimentions), so you should be able to make a preprocess script that can easily swap between encoder model.

Mortimerp9 reviewed (changes requested):
The HF dataset I put in my example is just an example, I suggest that you put this as a parameter so that this process can be applied on different datasets if we want to try different things (or look at a combination). Same goes for output file names, take them as input. They could have a default naming based on the input dataset.

You are loading the whole dataset and the whole embedding in memory, this will not scale to larger datasets, or image datasets. I suggest that you process the dataset in chunks/shards and save in split files on disk. It's unclear if UMAP would support processing shards, so maybe for that you still need to the whole embedding in memory.

This PR is a good start, I think that the sharding and multiple datasets together can be done in follow up PRs. But I think that you can already have parameters and better script organization in this PR.

Mortimerp9 commented:
some details on umap: https://umap-learn.readthedocs.io/en/latest/how_umap_works.html
for details about optimization for large datasets, see: https://github.com/lmcinnes/umap/issues/125 and the other issues linked at the bottom.

adaorachi committed:
refactor feature structure & explore scaling options

adaorachi commented:
@Mortimerp9 Thank you for your thorough review and feedback. Your insights were invaluable to my progress on this project. I have completed the changes requested.

### Changes Made:
- Restructured features into modules to improve comprehension.
- Addressed scaling issues with datasets and embeddings:
    - Identified limitations with Imagebind (it cannot embed a single point), so I opted to use sonar-space only for now.
    - Proposed integrating AutoImageProcessor from the transformer package for image embedding capability.
- Implemented argparse for passing arguments.
- Investigated UMAP performance enhancements:
    - Reviewed recommended resources; found potential scalability [here](https://umap-learn.readthedocs.io/en/latest/benchmarking.html)
    - Conducted benchmarking experiment:
       - Generated random data (1M points, 1024 dims).
       - Executed UMAP transform_fit with parameters {n_neighbors=15, n_components=2, min_dist=0.99}.
       - Completed in approximately 1762.3 seconds (~30 minutes) on device with specs (3.5 GHz Dual-Core Intel Core i7).

Please review the changes and provide feedback as necessary.

Mortimerp9 commented on backend/multimodalexplorer/api/endpoints/embeddings.py:
```diff
@@ -1,27 +1,21 @@
 
 from fastapi import APIRouter
 
+from ...functions.reduce_embed_dims import ReduceEmbedDims
```
it's better to use absolute imports, if you decide to move something around or rename a package, it becomes hard to find where the imports are used if the project uses relative imports.

Mortimerp9 commented on backend/multimodalexplorer/functions/create_faiss_index.py:
```diff
@@ -0,0 +1,86 @@
+
+        index = faiss.index_factory(vector_dims, "OPQ64,IVF1024,PQ64")
+
+        if DEVICE == "gpu":
```
I think that the device will be "cuda:0" and not "gpu", but it might be something else too, maybe use `torch.cuda.is_available()` isntead

Mortimerp9 commented on backend/multimodalexplorer/functions/create_faiss_index.py:
```diff
@@ -0,0 +1,86 @@
+        try:
+            self.create_index()
+        except Exception as e:
+            logger.exception(f"An error occurred: {e}")
```
let's be specific
```suggestion
            logger.exception("couldn't create faiss index", exc_info=e)
```

Mortimerp9 commented on backend/multimodalexplorer/functions/create_faiss_index.py:
```diff
@@ -0,0 +1,86 @@
+        file_path = get_file_path(file_name, f"../{self.save_index_dirname}")
+
+        faiss.write_index(index, file_path)
+        logger.info(f"Created Faiss index for embeddings - {index.ntotal}")
```
this is usually pretty slow, so maybe add other log message along the way to know the progress

Mortimerp9 commented on backend/multimodalexplorer/functions/helpers.py:
```diff
@@ -0,0 +1,42 @@
+def get_file_path(file_name, dir_name="../db"):
+    folder_path = os.path.join(os.path.dirname(__file__), dir_name)
+    os.makedirs(folder_path, exist_ok=True)
+    return os.path.join(folder_path, file_name)
```
I recommend that you use `pathlib.Path` for path manipulation and creation, it's a bit nicer than all the os bits.

But generally, I think that it's not a great idea to rely on relative path to a python project. The pre-processing script should take an `output_dir` path and that's where the artifacts should be saved. The fastapi server should also be launched with an `artifact_dir` path argument and load the data from there.

Mortimerp9 commented on backend/multimodalexplorer/functions/process_dataset.py:
```diff
@@ -0,0 +1,264 @@
+logger = logging.getLogger(__name__)
+
+# Determine the device (CPU or GPU) for torch operations
+DEVICE = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
```
let's not have a global constant in every file. Maybe add a `get_device` utils somewhere and use that when needed.

Mortimerp9 commented on backend/multimodalexplorer/functions/process_dataset.py:
```diff
@@ -0,0 +1,264 @@
+        self.chunk_size = chunk_size
+        self.embed_save_dir_name = embed_save_dir_name
+        self.data_save_dir_name = data_save_dir_name
+        self.device = DEVICE
```
i.e. here, you don't really need the global constant when you then add it to the instance properties.

Mortimerp9 commented on backend/multimodalexplorer/functions/process_dataset.py:
```diff
@@ -0,0 +1,264 @@
+                "Lengths of dataset_types and dataset_names must be equal."
+            )
+
+        valid_types = ["text", "image", "audio", "video"]
```
you can use enums in python: https://docs.python.org/3/library/enum.html

also, use types (from `import typing`) in your arguments etc. This will make this clearer.

Mortimerp9 commented on backend/multimodalexplorer/functions/process_dataset.py:
```diff
@@ -0,0 +1,264 @@
+
+            loaded_datasets[dataset_type] = loaded_dataset
+
+        return loaded_datasets
```
you might want to document in a docstring the expectations expressed in this bit of code.

Mortimerp9 commented on backend/multimodalexplorer/functions/process_dataset.py:
```diff
@@ -0,0 +1,264 @@
+            dataset.iter(batch_size=self.batch_size), desc=f"Embedding {dataset_type}"
+        ):
+            data = self._extract_data_from_batch(batch)
+            embeddings = data2vec_model.predict(data, source_lang="eng_Latn")
```
sonar supports different languages (that can be embedded in the same space), it might be good to take the lang as a parameter somewhere (maybe as a spec to the dataset list?).

Mortimerp9 commented on backend/multimodalexplorer/functions/process_dataset.py:
```diff
@@ -0,0 +1,264 @@
+            "audio": SpeechToEmbeddingModelPipeline(
+                encoder="sonar_speech_encoder_eng", device=self.device
+            ),
+        }[dataset_type]
```
these models are heavyish and slow to load, you might want to only load them if they are really needed.

Mortimerp9 commented on backend/multimodalexplorer/functions/reduce_embed_dims.py:
```diff
@@ -0,0 +1,78 @@
+        file_path = self._get_emb_path()
+
+        with open(file_path, "wb") as file:
+            pickle.dump(embeddings_umap, file)
```
this is a numpy array, use https://numpy.org/doc/stable/reference/generated/numpy.save.html instead of pickling, it will be easier to port between versions of python and faster to read/write in theory.

Mortimerp9 commented on backend/multimodalexplorer/functions/reduce_embed_dims.py:
```diff
@@ -0,0 +1,78 @@
+
+        try:
+            with open(file_path, "rb") as file:
+                embeddings = pickle.load(file)
```
see above

Mortimerp9 commented on backend/multimodalexplorer/functions/search_faiss_index.py:
```diff
@@ -0,0 +1,183 @@
+        Validate the provided arguments.
+        """
+        # Ensure dataset types are valid
+        valid_types = ["text", "image", "audio", "video"]
```
see comments about this. You shouldn't have this code duplicated, using an ENUM would help here.

Mortimerp9 commented on backend/multimodalexplorer/functions/search_faiss_index.py:
```diff
@@ -0,0 +1,183 @@
+            np.ndarray: Query_embedding as a NumPy array.
+        """
+
+        if not self.search_data:
+            raise ValueError("Search data is empty.")
+
```
you should load the index once in memory and then perform searches on it. the search_index function should have a search query parameter.

Mortimerp9 commented on backend/multimodalexplorer/functions/search_faiss_index.py:
```diff
@@ -0,0 +1,183 @@
+        if not self.search_data:
+            raise ValueError("Search data is empty.")
+
+        data2vec_model = {
+            "text": TextToEmbeddingModelPipeline(
+                encoder="text_sonar_basic_encoder",
+                tokenizer="text_sonar_basic_encoder",
+                device=self.device,
+            ),
+            "audio": SpeechToEmbeddingModelPipeline(
+                encoder="sonar_speech_encoder_eng", device=self.device
+            ),
+        }[self.search_data_type]
```
you should only load this once, it's big and slow. Load it in the init, or lazy load the right encoder pipeline depending on the query type.

Mortimerp9 commented on backend/multimodalexplorer/functions/search_faiss_index.py:
```diff
@@ -0,0 +1,183 @@
+        index = self._load_index_from_file()
+
+        query_embedding = self.process_search_query()
+        k = 5
```
k is the number of neighbours you are searching for, you might want to pass that as an argument to the search.

Mortimerp9 commented on backend/multimodalexplorer/functions/search_faiss_index.py:
```diff
@@ -0,0 +1,183 @@
+
+        raw_data = pd.read_csv(file_path, sep="\t")
+
+        list_searches = [raw_data.iloc[idx]["data"] for idx in I]
+        return list_searches
```
`D` are the embeddings, you don't really care for them, `I` is the id in the index, which is really the int index given by the order of insertion in the index. You might want to make sure that you can map this to the actual data point that it represents in your db (to retrieve the text/image/sound for display)

Mortimerp9 replied:
you don't have to fix this in this PR, but you'll need to have some consistency here when hooking this to the UI

Mortimerp9 reviewed (approved):
there are a few nits here and there plus some follow up changes that you'll probably need to do in future PRs

adaorachi committed:
implement types & lazy-load encoder pipelines

adaorachi committed:
cleanup create_faiss_index.py

fix isort lint error

fix isort lint error

fix black lint errors

fix isort lint error

Mortimerp9 reviewed (approved):
great, let's land this, looking forward for the backend<->frontend connection

Mortimerp9 commented:
try to run `isort --profile=black` and then `black`. You might need to upgrade your isort version to get the profile argument.

adaorachi committed:
fix black/isort linter errors

adaorachi committed:
add pre-commit hook for isort

adaorachi committed:
add pre-commit hook for isort

Mortimerp9 committed:
isort

Mortimerp9 committed:
setup poetry in gh action

Mortimerp9 committed:
setup poetry only in backend

Mortimerp9 committed:
fiix workflow

Mortimerp9 committed:
fiix workflow

Mortimerp9 committed:
fiix workflow

Mortimerp9 committed:
fiix workflow

Mortimerp9 committed:
fiix workflow

Mortimerp9 committed:
isort again

Mortimerp9 committed:
disable tests in workflow

Mortimerp9 commented:
ok, I've fixed the lint workflow, make sure to use the poetry installed isort (`poetry install` then `poetry run isort`) as this makes sure the versions are correct.
